sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/BlobDecryptionPolicy.java

```diff
@@ -151,67 +170,120 @@ Flux<ByteBuffer> decryptBlob(String encryptedDataString, Flux<ByteBuffer> encryp
+                                return plaintextByteBuffer;
+                            });
+                        case ENCRYPTION_PROTOCOL_V2:
+                            // Buffer an exact region with the nonce and tag
+                            BufferStagingArea stagingArea =
+                                new BufferStagingArea(GCM_ENCRYPTION_REGION_LENGTH + TAG_LENGTH + NONCE_LENGTH,
+                                GCM_ENCRYPTION_REGION_LENGTH + TAG_LENGTH + NONCE_LENGTH);
+
+                            return UploadUtils.chunkSource(encryptedFlux,
+                                    new com.azure.storage.common.ParallelTransferOptions()
+                                        .setBlockSizeLong((long) GCM_ENCRYPTION_REGION_LENGTH
+                                    + TAG_LENGTH + NONCE_LENGTH))
+                                .flatMapSequential(stagingArea::write)
+                                    .concatWith(Flux.defer(stagingArea::flush))
+                                    .flatMapSequential(aggregator -> {
+                                        // Get the IV out of the beginning of the aggregator
+                                        byte[] gmcIv = aggregator.getFirstNBytes(NONCE_LENGTH);
+
+                                        Cipher gmcCipher;
+                                        try {
+                                            gmcCipher = getCipher(contentEncryptionKey, encryptionData, gmcIv, false);
+                                        } catch (InvalidKeyException e) {
+                                            throw LOGGER.logExceptionAsError(Exceptions.propagate(e));
+                                        }
+
+                                        ByteBuffer decryptedRegion = ByteBuffer.allocate(GCM_ENCRYPTION_REGION_LENGTH);
+                                        return aggregator.asFlux()
+                                            .map(buffer -> {
+                                                // Write into the preallocated buffer and always return this buffer.
+                                                try {
+                                                    gmcCipher.update(buffer, decryptedRegion);
+                                                } catch (ShortBufferException e) {
+                                                    throw LOGGER.logExceptionAsError(Exceptions.propagate(e));
+                                                }
+                                                return decryptedRegion;
+                                            }).then(Mono.fromCallable(() -> {
+                                                // We have already written all the data to the cipher. Passing in a final
+                                                // empty buffer allows us to force completion and return the filled buffer.
+                                                gmcCipher.doFinal(ByteBuffer.allocate(0), decryptedRegion);
```

Comment:
Any reason this can't be

```java
ByteBuffer decryptedRegion = ByteBuffer.wrap(gcmCipher.doFinal());
decryptedRegion.flip();
return decryptedRegion;
```

This way we don't need to pre-allocate a ByteBuffer before decryption begins

Reply:
From what I remember of my experiments of how the cipher behaves, it does sometimes return data on update and not just on doFinal, so we need the preallocation to hold onto all the data along the way. The Cipher class is frustratingly vague about behavior and doesn't really give details about behavior under specific algorithms, so I can experiment with this and double check.